### DISCUSSION TITLE
GUI - how are the menus created? (adding addons to menus)

### DISCUSSION THREAD

**Discussion by CarlosGrohmann** [OSGeo/grass#1981] [Ideas]
Hello everyone

I was wondering, how are the menus for the GUI created? From a static file at compilation? Or dynamically at startup?
This is because I was thinking that we should have an easy way to have the installed addons available via the GUI menus.
The way QGIS does it, IMO is a bit messy: some plugins go to the "plugin" menu, while others create their own top-level menu (like MMQGIS)

If the GUI menus are (or can be) created dynamically, we could have it in two ways: adding the module to the "parent" top-level menu (raster, vector, etc), or to a "Extensions" menu (BTW, is there a discussion on "addon x extension"?)

Maybe every addon could have an XML (or similar) file indicating to which menu it should be added? 

I think this is important because many new users feel a bit lost when they install an addon but can't find it on the GUI afterward.

**Comment by wenzeslaus** [comment]
For the current system, see the [gui/wxpython/xml](https://github.com/OSGeo/grass/tree/main/gui/wxpython/xml) in the source code and the _Modules_ tab:

![Screenshot from 2021-11-16 14-01-47](https://user-images.githubusercontent.com/5449060/142048841-e84e5c66-9a83-4f19-9814-a869d8a936b3.png)

...but it could be better. Ideas welcome.

**Reply by CarlosGrohmann** [reply]
Thanks Anna!

If I understand it correctly, the GUI menus are build only during compilation? Or I could run `update_menudata.py` in an existing installation? If so, would it be possible to have a top-level "addons" menu? 

A little off-topic: what about the order of the items in the menus? how is it defined? (I've always wondered why r.mask is not above the "raster buffers" sub-menu...)

**Comment by petrasovaa** [comment]
> If I understand it correctly, the GUI menus are build only during compilation? Or I could run `update_menudata.py` in an existing installation? If so, would it be possible to have a top-level "addons" menu?

You can customize the menu, see if this would be helpful:
https://grass.osgeo.org/grass80/manuals/wxGUI.toolboxes.html

It may be a good idea to have the addons in the main menu, I think the concern was it would be too wide. That won't be an issue with a future single window layout though.
> 
> A little off-topic: what about the order of the items in the menus? how is it defined? (I've always wondered why r.mask is not above the "raster buffers" sub-menu...)
If you find the order weird or missing modules, it would be best to create a PR.
